toolchain/check/merge.cpp

```diff
@@ -223,15 +223,30 @@ static auto CheckRedeclParam(Context& context, bool is_implicit_param,
 
   auto new_param_pattern = context.insts().Get(new_param_pattern_id);
   auto prev_param_pattern = context.insts().Get(prev_param_pattern_id);
-  if (new_param_pattern.kind() != prev_param_pattern.kind() ||
-      !context.types().AreEqualAcrossDeclarations(
-          new_param_pattern.type_id(),
-          SemIR::GetTypeInSpecific(context.sem_ir(), prev_specific_id,
-                                   prev_param_pattern.type_id()))) {
+  if (new_param_pattern.kind() != prev_param_pattern.kind()) {
     emit_diagnostic();
     return false;
   }
 
+  auto prev_param_type_id = SemIR::GetTypeInSpecific(
+      context.sem_ir(), prev_specific_id, prev_param_pattern.type_id());
+  if (!context.types().AreEqualAcrossDeclarations(new_param_pattern.type_id(),
+                                                  prev_param_type_id)) {
+    if (!diagnose) {
+      return false;
+    }
+    CARBON_DIAGNOSTIC(
+        RedeclParamDiffersType, Error,
+        "redeclaration differs at {0:implicit |}parameter {1} type {2} != {3}",
```

Comment:
I think this could be clearer about which type is which. For example:
```suggestion
        "type {3} of {0:implicit |}parameter {1} in redeclaration differs from previous parameter type {2}",
```